README.md
# Emacs Configuration.

I use emacs trunk for my daily tasks, I update it often, so my configuration
might not work with earlier version of emacs.

## Installation

If you already have .emacs.d directory do:

	mv ~/.emacs.d ~/.my.emacs.d

then:

	git clone https://github.com/xmonk/emacs.d.git ~/.emacs.d


CHANGELOG.md
### Fix issue with url-insert-buffer-content (HEAD -> master, origin/master, origin/HEAD)
>Fri, 20 Nov 2015 22:45:27 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)

The function url-insert-buffer-content is expected to be bound, but as
it's not autoloaded package.el doesn't update or connects. This code
make sure that if it's not bound the librar url-handlers is loaded
for the function to be bound.



### Enable custom theme.
>Sun, 8 Nov 2015 19:49:00 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Add box to inactive modeline.
>Sun, 8 Nov 2015 19:48:32 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Fix formatting and add doc strings.
>Sun, 8 Nov 2015 19:47:50 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Set region to a sane color.
>Sun, 8 Nov 2015 10:08:50 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Fix save-place.
>Tue, 3 Nov 2015 12:28:54 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Updated to use use-package for save-place
>Sat, 31 Oct 2015 21:51:46 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Refactored (jj/print-date) into (jj/insert-date) (jj/insert-datetime)
>Sat, 31 Oct 2015 13:37:03 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### New packages
>Thu, 29 Oct 2015 16:47:33 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Add orgmode elpa
>Thu, 29 Oct 2015 16:47:20 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Fix a few issues, changed foreground color.
>Tue, 27 Oct 2015 16:24:08 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Add go-rename package.
>Tue, 27 Oct 2015 15:30:21 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Fix typo.
>Tue, 27 Oct 2015 15:30:05 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### More formatting.
>Tue, 27 Oct 2015 15:29:49 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Add comment to enable a dark theme.
>Tue, 27 Oct 2015 15:29:20 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Formatting.
>Tue, 27 Oct 2015 15:27:59 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Use go-rename melpa package.
>Tue, 27 Oct 2015 15:26:16 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Reverting to default modeline.
>Sat, 24 Oct 2015 18:51:32 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Experimenting with fonts.
>Sat, 24 Oct 2015 18:48:13 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Update packag-selected-packages.
>Sat, 24 Oct 2015 18:47:09 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Disabled helm
>Sat, 24 Oct 2015 18:42:06 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Change keybinding for eshell
>Sat, 24 Oct 2015 18:41:46 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Fix issue with utop config not loading.
>Sat, 24 Oct 2015 18:41:23 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### set cursor color
>Fri, 23 Oct 2015 18:16:59 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Update header.
>Thu, 22 Oct 2015 18:55:43 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Add file header.
>Thu, 22 Oct 2015 18:55:25 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Disable use-package verbose
>Thu, 22 Oct 2015 17:07:00 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Trying a dark background for a change.
>Thu, 22 Oct 2015 17:06:41 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Ensure packages are installed.
>Thu, 22 Oct 2015 17:06:07 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Refactored codesearch function into a macro.
>Thu, 22 Oct 2015 17:04:23 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)

By using a macro instead of several functions, I can create different
functions using different index files, making search even more useful.



### removed semantic
>Wed, 21 Oct 2015 18:01:36 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### rempped swiper keys
>Wed, 21 Oct 2015 18:01:25 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### enable semantic
>Wed, 21 Oct 2015 17:35:23 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Add new items to gitignore.
>Tue, 20 Oct 2015 08:29:56 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Enable ocaml setup.
>Tue, 20 Oct 2015 08:26:54 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Add new packages.
>Tue, 20 Oct 2015 08:26:26 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Cleaned up helm configuration.
>Tue, 20 Oct 2015 08:25:39 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)




### Update swiper configuration.
>Tue, 20 Oct 2015 08:25:10 -0400

>Author: Juan Fuentes (jj@xmonk.org)

>Commiter: Juan Fuentes (jj@xmonk.org)





init.el
;;; init.el --- Emacs configuration.
;;
;; Filename: init.el
;; Description:
;; Author: Juan Fuentes
;; Maintainer:
;; Created: Sat Jun 28 18:48:44 2008
;; Version:
;; Package-Requires: ()
;; Last-Updated: Sat Apr 18 14:22:27 2015 (-0400)
;;           By:
;;     Update #: 92
;; URL:
;; Doc URL:
;; Keywords:
;; Compatibility:
;;
;;
;;; Commentary:
;;
;;
;;; Change Log:
;;
;;
;;; Code:

(require 'debug)
(setq debug-on-error nil)

(defvar *site-lisp* (concat user-emacs-directory "site-lisp/")
  "Location of configuration files to be loaded at start up.")

(defvar *vendor* (concat user-emacs-directory "vendor/")
  "Location of third-party files to be loaded at start up.")

(defvar running-apple (string-match "apple" system-configuration))

;; 20MB let's see how well it works.
(setq gc-cons-threshold 20000000)

(unless window-system
  (menu-bar-mode -1))

(dolist (mode '(scroll-bar-mode tool-bar-mode))
  (if (fboundp mode) (funcall mode -1)))

(when running-apple
  (if (boundp 'mac-option-modifier)
      (setq mac-option-modifier 'meta))
  (setq mac-allow-anti-aliasing t))

;; set faces
;; Uncomment following two lines and comment the third for dark background
;; (set-face-attribute 'default nil :background "black" :foreground "#D0D0D0" :font "Lucida Grande Mono" :height 120 :slant 'normal :weight 'normal)
;; (set-face-attribute 'cursor nil :background "#D0D0D0" :foreground "#D0D0D0")
(set-face-attribute 'default nil :font "Lucida Grande Mono" :height 120 :slant 'normal :weight 'normal)
(set-face-attribute 'mode-line nil :box nil :font "Lucida Grande" :height 120 :slant 'normal :weight 'normal)
(set-face-attribute 'font-lock-comment-face nil :font "Lucida Grande" :height 130 :slant 'italic :weight 'normal)
(set-face-attribute 'font-lock-doc-face nil :font "Lucida Grande" :height 130 :slant 'normal :weight 'normal)
(set-face-attribute 'font-lock-function-name-face nil :font "Lucida Grande" :height 130 :slant 'normal :weight 'normal)

;; In os x there is an issue with tramp and TMPDIR, this is a work around.
(if running-apple
    (setenv "TMPDIR" "/tmp"))

;; Load customization's
(cond ((eql system-type 'darwin)
       (setq custom-file (concat user-emacs-directory "osx-custom.el")))
      ((eql system-type 'berkeley-unix)
       (setq custom-file (concat user-emacs-directory "obsd-custom.el")))
      ((eql system-type 'gnu/linux)
       (setq custom-file (concat user-emacs-directory "linux-custom.el"))))

(if (file-exists-p custom-file)
    (load-file custom-file)
  (message "ERROR: No custom file found or specified"))

;; set exec-path
(when (eq window-system 'ns)
  (let ((path (shell-command-to-string "$SHELL -cl \"printf %s \\\"\\\$PATH\\\"\"")))
    (setenv "PATH" path)
    (setq exec-path (split-string path path-separator))))

;;; Server
(require 'server nil t)
(when (and (>= emacs-major-version 23) (not (server-running-p)))
  (server-start))

;;; This fixes the issue of 'url-insert-buffer-content not being autoloaded.
(unless (fboundp 'url-insert-buffer-contents)
  (load-library "url-handlers"))

;; package
(autoload 'package "package" nil t)

(setq package-archives
      '(("gnu"   . "https://elpa.gnu.org/packages/")
        ("melpa" . "https://melpa.org/packages/")
	("org" . "http://orgmode.org/elpa/")))

(package-initialize)

(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))

(require 'use-package)
(defvar use-package-verbose t)

(add-to-list 'load-path (expand-file-name *site-lisp*))
(use-package jj-theme)
(use-package defuns)
(use-package global)
(use-package keymaps)
(use-package helm-conf)
(use-package elisp-conf)
(use-package lisp-conf)
(use-package c-conf)
(use-package go-conf)
(use-package org-conf)
(use-package magit-conf)
(use-package py-conf)
(use-package ocaml-conf)
(use-package java-conf)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; init.el ends here
